src/compiler/checker.ts

```diff
@@ -5351,17 +5351,39 @@ module ts {
                 }
                 // Target type is type of prototype property
                 let prototypeProperty = getPropertyOfType(rightType, "prototype");
-                if (!prototypeProperty) {
-                    return type;
+                if (prototypeProperty) {
+                    let targetType = getTypeOfSymbol(prototypeProperty);
+                    if (targetType !== anyType) {
+                        // Narrow to target type if it is a subtype of current type
+                        if (isTypeSubtypeOf(targetType, type)) {
+                            return targetType;
+                        }
+                        // If current type is a union type, remove all constituents that aren't subtypes of target type
+                        if (type.flags & TypeFlags.Union) {
+                            return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, targetType)));
+                        }
+                    }
                 }
-                let targetType = getTypeOfSymbol(prototypeProperty);
-                // Narrow to target type if it is a subtype of current type
-                if (isTypeSubtypeOf(targetType, type)) {
-                    return targetType;
+                // Target type is type of constructor signiture
+                let constructSignitures: Signature[];
+                if (rightType.flags & TypeFlags.Interface) {
+                    constructSignitures = (<InterfaceTypeWithDeclaredMembers>rightType).declaredConstructSignatures;
                 }
-                // If current type is a union type, remove all constituents that aren't subtypes of target type
-                if (type.flags & TypeFlags.Union) {
-                    return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, targetType)));
+                if (rightType.flags & TypeFlags.Anonymous) {
+                    constructSignitures = (<ResolvedType>rightType).constructSignatures;
+                }
+                if (constructSignitures) {
+                    let instanceType = getUnionType(map(constructSignitures, constructSignature => {
+                        if (constructSignature.typeParameters && constructSignature.typeParameters.length !== 0) {
+                            constructSignature = instantiateSignature(constructSignature, createTypeMapper(constructSignature.typeParameters, map(constructSignature.typeParameters, _ => anyType)), true)
+                        }
+                        return constructSignature.resolvedReturnType;
+                    }));
+                    // Pickup type from union types
+                    if (type.flags & TypeFlags.Union) {
+                        return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, instanceType)));

```

Comment:
Since we do this twice, maybe factor it out into a function called `getUnionTypeOfSubtypeConstituents`